src/imitation/data/types.py

```diff
@@ -22,7 +29,267 @@
+        """Returns the array for the given key, or raises KeyError."""
+        return self._d[key]
+
+    def unwrap(self) -> Dict[str, np.ndarray]:
+        """Returns a copy of the underlying dictionary (arrays are not copied)."""
+        return {k: v for k, v in self._d.items()}
+
+    def map_arrays(self, fn: Callable[[np.ndarray], np.ndarray]) -> "DictObs":
+        """Returns a new DictObs with `fn` applied to every array."""
+        return self.__class__({k: fn(v) for k, v in self.items()})
+
+    @staticmethod
+    def _unravel(dictobs_list: Iterable["DictObs"]) -> Dict[str, List[np.ndarray]]:
+        """Converts a list of DictObs into a dictionary of lists of arrays."""
+        it1, it2 = itertools.tee(dictobs_list)
+        # assert all have same keys
+        key_set = set(frozenset(obs.keys()) for obs in it1)
+        if len(key_set) == 0:
+            raise ValueError("Empty list of DictObs")
+        if not len(key_set) == 1:
+            raise ValueError(f"Inconsistent keys: {key_set}")
+
+        unraveled: Dict[str, List[np.ndarray]] = collections.defaultdict(list)
+        for do in it2:
+            for k, array in do._d.items():
+                unraveled[k].append(array)
+        return unraveled
+
+    @classmethod
+    def stack(cls, dictobs_list: Iterable["DictObs"], axis=0) -> "DictObs":
+        """Returns a single dictobs stacking the arrays by key."""
+        return cls(
+            {
+                k: np.stack(arr_list, axis=axis)
+                for k, arr_list in cls._unravel(dictobs_list).items()
+            },
+        )
+
+    @classmethod
+    def concatenate(cls, dictobs_list: Iterable["DictObs"], axis=0) -> "DictObs":
+        """Returns a single dictobs concatenating the arrays by key."""
+        return cls(
+            {
+                k: np.concatenate(arr_list, axis=axis)
+                for k, arr_list in cls._unravel(dictobs_list).items()
+            },
+        )
+
+
+# DicObs utilities
+
+
+Observation = Union[np.ndarray, DictObs]
+ObsVar = TypeVar("ObsVar", np.ndarray, DictObs)
+
+
+def assert_not_dictobs(x: Observation) -> np.ndarray:
+    """Typeguard to assert `x` is an array, not a DictObs."""
+    if isinstance(x, DictObs):
+        assert False, "Dictionary observations are not supported here."
+    return x
+
+
+def concatenate_maybe_dictobs(arrs: List[ObsVar]) -> ObsVar:
+    """Concatenates a list of observations appropriately (depending on type)."""
+    assert len(arrs) > 0
+    if isinstance(arrs[0], DictObs):
+        return DictObs.concatenate(arrs)
+    else:
+        return np.concatenate(arrs)
+
+
+def stack_maybe_dictobs(arrs: List[ObsVar]) -> ObsVar:
+    """Stacks a list of observations appropriately (depending on type)."""
+    assert len(arrs) > 0
+    if isinstance(arrs[0], DictObs):
+        return DictObs.stack(arrs)
+    else:
+        return np.stack(arrs)
+
+
+# the following overloads have a type error as a DictObs matches both definitions, but
```

Comment:
Sad. Could maybe have overloads for each expected value of T (`np.ndarray`, `th.Tensor`, `int`) and just treat the others as type errors? Unsure how strict we want to be for `maybe_unwrap_dictobs`.